server/rest/src/main/java/org/infinispan/rest/stream/CacheChunkedStream.java

```diff
@@ -0,0 +1,205 @@
+   @Override
+   public long progress() {
+      return progress;
+   }
+
+   abstract static class ByteBufSubscriber<T> extends DefaultSubscriber<T> {
+      protected final ChannelHandlerContext ctx;
+      protected final ByteBufAllocator allocator;
+
+      private volatile boolean hasReturnedNull;
+      private boolean completed;
+      private boolean firstEntry = true;
+
+      private final AtomicReference<ByteBuf> readyBuffer = new AtomicReference<>();
+      private final AtomicReference<ByteBuf> pendingBuffer = new AtomicReference<>();
+
+      protected ByteBufSubscriber(ChannelHandlerContext ctx, ByteBufAllocator allocator) {
+         this.ctx = Objects.requireNonNull(ctx);
+         this.allocator = Objects.requireNonNull(allocator);
+      }
+
+      protected ByteBuf newByteBuf(int size) {
+         // Always need array to be 1 bigger than actual size for the ending bracket
+         int allocatedSize = Math.max(size + 1, CHUNK_SIZE);
+         return allocator.buffer(allocatedSize, allocatedSize);
+      }
+
+      @Override
+      protected void onStart() {
+         ByteBuf buf = newByteBuf(0);
+         buf.writeByte('[');
+         pendingBuffer.set(buf);
+         request(1);
+      }
+
+      @Override
+      public void onNext(T t) {
+         ByteBuf pendingBuf = pendingBuffer.get();
+         if (!firstEntry) {
+            pendingBuf.writeByte(',');
+         } else {
+            firstEntry = false;
+         }
+         ByteBuf newBuf = writeItem(t, pendingBuf);
+         if (newBuf != null) {
+            assert pendingBuf.writableBytes() == 0;
+            // It is technically possible that a single item may be larger than remaining pending and the new buffer
+            // In that case the implementation must return a larger ByteBuf that we provided, however it must also
+            // have at least a single writeable byte in case this is the last value and we need to append ] to it
+            assert newBuf.writableBytes() > 0;
+            pendingBuffer.set(newBuf);
+            readyBuffer.set(pendingBuf);
+
+            notifyReadyBuffer();
+         } else {
+            assert pendingBuf.writableBytes() > 0;
+            request(1);
+         }
+      }
+
+      // Implementation must write the bytes for the item to the pending buffer first. If the pending buffer does not
+      // contain enough bytes, the newByteBuf method should be invoked with how many bytes it requires and then write
+      // the same number of bytes into the returned ByteBuf and return the ByteBuf from this method
+      abstract ByteBuf writeItem(T item, ByteBuf pending);
+
+      @Override
+      public void onError(Throwable t) {
+         logger.error("Error encountered while streaming cache chunk", t);
+         // TODO: is there a way to propagate?
+      }
+
+      @Override
+      public void onComplete() {
+         ByteBuf buf = pendingBuffer.getAndSet(null);
+         buf.writeByte(']');
+         readyBuffer.set(buf);
```

Comment:
Ah, it isn't. I thought the spec worked differently, in that an onComplete couldn't happen without an outstanding request. According to 2.9 this can happen though https://github.com/reactive-streams/reactive-streams-jvm/blob/v1.0.4/README.md#specification